File: packages/native/engine-native/test/fixtures.ts
import type { QuaNativeHostApi, QuaNativeHostInfo, TargetBundleManifest } from '@quajs/native-contracts'
import {
  createNativeCapabilityManifestHash,
  getTargetCorePluginFamily,
  getTargetCoreResolverId,
  NATIVE_TARGET_BOOTSTRAP,
} from '@quajs/native-contracts'
import { vi } from 'vitest'

export const CAPABILITIES: QuaNativeHostInfo['renderer']['capabilities'] = [
  {
    id: 'native-wgpu.stage-layout@1',
    target: 'native',
    version: '1.0.0',
    ownerPackage: '@quajs/native-renderer',
    projectionKeys: ['QuaViewProjection.layout'],
    qssFeatures: ['safe-area', 'logical-stage'],
    quiComponents: ['Stage'],
    fallback: 'reject-package',
  },
  {
    id: 'native-wgpu.scene-transition@1',
    target: 'native',
    version: '1.0.0',
    ownerPackage: '@quajs/native-renderer',
    projectionKeys: ['view.sceneTransition'],
    intentEvents: ['scene/ready'],
    fallback: 'reject-package',
  },
  {
    id: 'native-wgpu.effects@1',
    target: 'native',
    version: '1.0.0',
    ownerPackage: '@quajs/native-renderer',
    projectionKeys: ['view.effects'],
    fallback: 'reject-package',
  },
  {
    id: 'native-wgpu.image@1',
    target: 'native',
    version: '1.0.0',
    ownerPackage: '@quajs/native-renderer',
    projectionKeys: ['background', 'characters', 'ui.image'],
    assetKinds: ['images', 'characters'],
    qssFeatures: ['object-fit', 'object-position', 'opacity'],
    quiComponents: ['Image'],
    fallback: 'render-empty',
  },
  {
    id: 'native-wgpu.video@1',
    target: 'native',
    version: '1.0.0',
    ownerPackage: '@quajs/native-renderer',
    projectionKeys: ['background.video'],
    assetKinds: ['video', 'images'],
    qssFeatures: ['object-fit', 'object-position', 'opacity'],
    fallback: 'warn-once',
  },
  {
    id: 'native-wgpu.text@1',
    target: 'native',
    version: '1.0.0',
    ownerPackage: '@quajs/native-renderer',
    projectionKeys: ['dialogue', 'ui.text'],
    assetKinds: ['fonts'],
    qssFeatures: [
      'font-family',
      'font-size',
      'font-style',
      'font-weight',
      'letter-spacing',
      'line-height',
      'text-align',
      'text-decoration',
      'text-overflow',
      'text-transform',
      'white-space',
      'color',
    ],
    quiComponents: ['Text', 'RichText'],
    fallback: 'warn-once',
  },
  {
    id: 'native-wgpu.ui.surface@1',
    target: 'native',
    version: '1.0.0',
    ownerPackage: '@quajs/native-renderer',
    projectionKeys: ['view.ui.overlays'],
    intentEvents: ['ui/intent', 'choice/select'],
    assetKinds: ['data', 'images', 'fonts', 'qui', 'qss', 'tokens'],
    qssFeatures: [
      'align-items',
      'background-color',
      'background-image',
      'background-position',
      'background-size',
      'border-color',
      'border-radius',
      'border-style',
      'border-width',
      'bottom',
      'box-sizing',
      'column-gap',
      'color',
      'display',
      'font-family',
      'font-size',
      'font-style',
      'font-weight',
      'gap',
      'inset',
      'letter-spacing',
      'height',
      'left',
      'line-height',
      'justify-content',
      'margin',
      'margin-bottom',
      'margin-left',
      'margin-right',
      'margin-top',
      'max-height',
      'max-width',
      'min-height',
      'min-width',
      'text-align',
      'text-decoration',
      'text-overflow',
      'text-transform',
      'object-fit',
      'object-position',
      'opacity',
      'overflow',
      'padding',
      'padding-bottom',
      'padding-left',
      'padding-right',
      'padding-top',
      'pointer-events',
      'position',
      'right',
      'row-gap',
      'top',
      'visibility',
      'white-space',
      'width',
      'z-index',
    ],
    quiComponents: [
      'Box',
      'Backdrop',
      'Button',
      'Column',
      'Divider',
      'Fragment',
      'Grid',
      'Layer',
      'Row',
      'Text',
      'RichText',
      'Image',
      'Panel',
      'SafeArea',
      'Scroll',
      'Spacer',
      'Stack',
    ],
    fallback: 'reject-package',
  },
  {
    id: 'native-wgpu.input.pointer@1',
    target: 'native',
    version: '1.0.0',
    ownerPackage: '@quajs/native-renderer',
    projectionKeys: ['view.choices', 'view.ui.overlays'],
    intentEvents: ['choice/select', 'ui/intent'],
    quiComponents: ['Backdrop', 'Box', 'Button', 'Panel'],
    fallback: 'reject-package',
  },
  {
    id: 'native-wgpu.input.text@1',
    target: 'native',
    version: '1.0.0',
    ownerPackage: '@quajs/native-renderer',
    projectionKeys: [],
    intentEvents: ['user/text_input'],
    fallback: 'reject-package',
  },
]

export const CAPABILITY_MANIFEST_HASH = createNativeCapabilityManifestHash(
  CAPABILITIES,
  payload => `test-sha256:${payload.length}`,
)

export function escapeRegExp(value: string): string {
  return value.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')
}

export function createHostInfo(version = '0.1.0'): QuaNativeHostInfo {
  return {
    app: {
      name: 'Native Fixture',
      bundleId: 'dev.quajs.native.fixture',
      version: '1.0.0',
      buildNumber: '100',
      profile: 'debug',
      platform: 'macos',
      arch: 'arm64',
    },
    renderer: {
      packageName: '@quajs/native-renderer',
      version,
      backend: 'wgpu',
      capabilityManifestHash: CAPABILITY_MANIFEST_HASH,
      capabilities: CAPABILITIES,
    },
    runtime: {
      quickjsVersion: '2025-04-26',
      nativeRuntimeVersion: '0.1.0',
      assetAdapterVersion: '0.1.0',
      storeAdapterVersion: '0.1.0',
    },
  }
}

export function createHost(hostInfo = createHostInfo()): QuaNativeHostApi {
  return {
    getHostInfo: vi.fn(async () => hostInfo),
    readAssetBytes: vi.fn(),
    readStorage: vi.fn(),
    writeStorage: vi.fn(),
    deleteStorage: vi.fn(),
    hashBytes: vi.fn(),
  }
}

export function createTestPipeline() {
  const listeners = new Map<string, Set<(context: any) => unknown>>()
  return {
    on: vi.fn((type: string, listener: (context: any) => unknown) => {
      const eventListeners = listeners.get(type) || new Set()
      eventListeners.add(listener)
      listeners.set(type, eventListeners)
    }),
    off: vi.fn((type: string, listener: (context: any) => unknown) => {
      listeners.get(type)?.delete(listener)
    }),
    emit: vi.fn(async (type: string, payload: unknown) => {
      for (const listener of listeners.get(type) || []) {
        await listener({
          event: {
            type,
            payload,
            timestamp: Date.now(),
            id: `${type}:test`,
          },
          handled: false,
          stopPropagation: false,
        })
      }
    }),
  }
}

export function createNativeTargetBundleManifest(
  overrides: Partial<TargetBundleManifest> = {},
): TargetBundleManifest {
  return {
    schemaVersion: 1,
    target: 'native',
    profile: 'debug',
    platform: 'macos',
    app: {
      bundleId: 'dev.quajs.native.fixture',
      version: '1.0.0',
      buildNumber: '100',
      icon: 'AppIcon.icns',
    },
    nativeRenderer: {
      packageName: '@quajs/native-renderer',
      version: '0.1.0',
      backend: 'wgpu',
      capabilityIds: CAPABILITIES.map(capability => capability.id),
      capabilityManifestHash: CAPABILITY_MANIFEST_HASH,
    },
    nativeRuntime: {
      quickjsVersion: '2025-04-26',
      nativeRuntimeVersion: '0.1.0',
      assetAdapterVersion: '0.1.0',
      storeAdapterVersion: '0.1.0',
    },
    targetCoreResolver: getTargetCoreResolverId('native'),
    selectedCorePluginFamily: getTargetCorePluginFamily('native'),
    selectedCoreAdapters: NATIVE_TARGET_BOOTSTRAP.coreAdapters,
    dependencies: [
      '@quajs/engine',
      '@quajs/pipeline',
      ...NATIVE_TARGET_BOOTSTRAP.coreAdapters,
    ],
    rendererEntries: [
      { specifier: '@quajs/native-renderer/builtin', target: 'native' },
    ],
    runtimePackages: [
      {
        id: 'runtime.chapter.native-ui',
        executableDependencies: ['@quajs/character'],
        rendererEntries: [
          { specifier: '@quajs/native-renderer/ui', target: 'native' },
        ],
      },
    ],
    ...overrides,
  }
}

export function createTrustContext(overrides: Record<string, unknown> = {}) {
  const runtimePackage = {
    id: 'runtime.chapter.native-ui',
    version: '1.0.0',
    scripts: [
      { id: 'opening', assetName: 'scripts/opening.js' },
    ],
    ...overrides,
  }
  return {
    package: runtimePackage,
    bundle: {
      packageId: runtimePackage.id,
      bundleName: runtimePackage.id,
      version: '1.0.0',
      bundleVersion: 1,
      hash: 'hash',
      priority: 0,
      loadedAt: 1,
      assetCount: 1,
      manifest: {
        assets: {},
      },
    },
  } as any
}

export function createModuleLoadContext(assetCodeByName: Record<string, string> = {}) {
  const calls: unknown[] = []
  const ctx = {
    assets: {
      async getAsset(type: string, name: string, options: unknown) {
        calls.push({ type, name, options })
        return {
          data: new TextEncoder().encode(assetCodeByName[name] || `export default "${name}"`),
        }
      },
    },
    bundle: {
      packageId: 'runtime.chapter.native-ui',
      bundleName: 'runtime.chapter.native-ui',
      version: '1.0.0',
      bundleVersion: 1,
      hash: 'hash',
      priority: 0,
      loadedAt: 1,
      assetCount: 1,
      manifest: {
        name: 'runtime.chapter.native-ui',
        version: 1,
        buildNumber: '100',
        created: '2026-06-24T00:00:00.000Z',
        assets: [],
      },
    },
    package: {
      id: 'runtime.chapter.native-ui',
      version: '1.0.0',
    },
    locale: 'ja-JP',
  }
  return { calls, ctx: ctx as any }
}

File: packages/native/crates/quajs_native_app/src/target_bundle/tests/fixtures.rs
use super::*;

pub(crate) fn unique_manifest_path(label: &str) -> std::path::PathBuf {
    std::env::temp_dir().join(format!(
        "quajs-native-target-bundle-{label}-{}-{}.json",
        std::process::id(),
        std::thread::current().name().unwrap_or("test")
    ))
}

pub(crate) fn native_manifest_json() -> serde_json::Value {
    json!({
        "target": "native",
        "profile": current_profile_value(),
        "platform": current_platform_value(),
        "app": {
            "bundleId": "dev.quajs.fixture",
            "version": "1.0.0",
            "buildNumber": "100",
            "icon": "AppIcon.icns"
        },
        "nativeRenderer": {
            "packageName": "@quajs/native-renderer",
            "version": "0.1.0",
            "backend": "wgpu",
            "capabilityIds": native_capability_ids(),
            "capabilityManifestHash": native_capability_manifest_hash()
        },
        "nativeRuntime": {
            "quickjsVersion": quajs_native_runtime::quickjs_runtime_version(),
            "nativeRuntimeVersion": env!("CARGO_PKG_VERSION"),
            "assetAdapterVersion": env!("CARGO_PKG_VERSION"),
            "storeAdapterVersion": env!("CARGO_PKG_VERSION")
        },
        "targetCoreResolver": "native-core-resolver",
        "selectedCorePluginFamily": "native-core",
        "selectedCoreAdapters": [
            "@quajs/engine-native/native-host",
            { "specifier": "@quajs/assets-native", "runtime": true },
            { "packageName": "@quajs/store-native" },
            "@quajs/native-contracts/bootstrap"
        ],
        "dependencies": [
            "@quajs/engine",
            "@quajs/pipeline",
            "quajs_wgpu_renderer::capabilities"
        ],
        "rendererEntries": [
            { "specifier": "@quajs/native-renderer/builtin", "target": "native" }
        ],
        "runtimePackages": [
            {
                "id": "runtime.chapter.native-ui",
                "executableDependencies": ["@quajs/character"],
                "rendererEntries": [
                    { "specifier": "@quajs/native-renderer/ui", "target": "native" }
                ]
            }
        ],
        "projectGraphs": [
            {
                "id": "native.startup-shell",
                "kind": "startup-shell",
                "references": [
                    "@quajs/engine",
                    "@quajs/character"
                ]
            },
            {
                "id": "native.release.macos.post-bundle",
                "kind": "post-bundle",
                "references": [
                    "@quajs/engine",
                    "@quajs/engine-native",
                    "@quajs/assets-native",
                    "@quajs/store-native",
                    "@quajs/native-contracts",
                    "quajs_wgpu_renderer::capabilities"
                ]
            }
        ]
    })
}

pub(crate) fn native_manifest() -> NativeTargetBundleManifest {
    NativeTargetBundleManifest {
        target: "native".to_string(),
        profile: current_profile_value().to_string(),
        platform: Some(current_platform_value().to_string()),
        app: Some(TargetBundleAppInfo {
            bundle_id: Some("dev.quajs.fixture".to_string()),
            version: Some("1.0.0".to_string()),
            build_number: Some("100".to_string()),
            icon: Some("AppIcon.icns".to_string()),
        }),
        native_renderer: Some(TargetBundleNativeRendererInfo {
            package_name: Some("@quajs/native-renderer".to_string()),
            version: Some("0.1.0".to_string()),
            backend: Some("wgpu".to_string()),
            backend_version: None,
            capability_ids: native_capability_ids(),
            capability_manifest_hash: Some(native_capability_manifest_hash()),
        }),
        native_runtime: Some(TargetBundleNativeRuntimeInfo {
            quickjs_version: Some(quajs_native_runtime::quickjs_runtime_version().to_string()),
            native_runtime_version: Some(env!("CARGO_PKG_VERSION").to_string()),
            asset_adapter_version: Some(env!("CARGO_PKG_VERSION").to_string()),
            store_adapter_version: Some(env!("CARGO_PKG_VERSION").to_string()),
        }),
        target_core_resolver: Some("native-core-resolver".to_string()),
        selected_core_plugin_family: "native-core".to_string(),
        selected_core_adapters: NATIVE_CORE_ADAPTERS
            .iter()
            .map(|specifier| TargetBundleReference::Specifier((*specifier).to_string()))
            .collect(),
        dependencies: vec![
            TargetBundleReference::Specifier("@quajs/engine".to_string()),
            TargetBundleReference::Specifier("@quajs/pipeline".to_string()),
            TargetBundleReference::Specifier("quajs_wgpu_renderer::capabilities".to_string()),
        ],
        renderer_entries: vec![TargetBundleReference::Object(TargetBundleReferenceObject {
            specifier: Some("@quajs/native-renderer/builtin".to_string()),
            package_name: None,
            target: Some("native".to_string()),
            plugin_id: None,
        })],
        runtime_packages: vec![RuntimePackageRecord {
            id: "runtime.chapter.native-ui".to_string(),
            executable_dependencies: vec![TargetBundleReference::Specifier(
                "@quajs/character".to_string(),
            )],
            renderer_entries: vec![TargetBundleReference::Object(TargetBundleReferenceObject {
                specifier: Some("@quajs/native-renderer/ui".to_string()),
                package_name: None,
                target: Some("native".to_string()),
                plugin_id: None,
            })],
        }],
        project_graphs: vec![
            TargetBundleProjectGraphRecord {
                id: "native.startup-shell".to_string(),
                kind: "startup-shell".to_string(),
                references: vec![
                    TargetBundleReference::Specifier("@quajs/engine".to_string()),
                    TargetBundleReference::Specifier("@quajs/character".to_string()),
                ],
            },
            TargetBundleProjectGraphRecord {
                id: "native.release.macos.post-bundle".to_string(),
                kind: "post-bundle".to_string(),
                references: vec![
                    TargetBundleReference::Specifier("@quajs/engine".to_string()),
                    TargetBundleReference::Specifier("@quajs/engine-native".to_string()),
                    TargetBundleReference::Specifier("@quajs/assets-native".to_string()),
                    TargetBundleReference::Specifier("@quajs/store-native".to_string()),
                    TargetBundleReference::Specifier("@quajs/native-contracts".to_string()),
                    TargetBundleReference::Specifier(
                        "quajs_wgpu_renderer::capabilities".to_string(),
                    ),
                ],
            },
        ],
    }
}

pub(crate) fn native_expectation() -> NativeStartupManifestExpectation {
    NativeStartupManifestExpectation {
        bundle_id: "dev.quajs.fixture".to_string(),
        version: "1.0.0".to_string(),
        build_number: "100".to_string(),
        profile: current_profile_value().to_string(),
        platform: current_platform_value().to_string(),
    }
}

fn current_profile_value() -> &'static str {
    profile_manifest_value(quajs_native_runtime::current_profile())
}

fn current_platform_value() -> &'static str {
    platform_manifest_value(quajs_native_runtime::current_platform())
}

fn native_capability_ids() -> Vec<String> {
    crate::startup::native_startup_renderer_capabilities()
        .into_iter()
        .map(|capability| capability.id)
        .collect()
}

fn native_capability_manifest_hash() -> String {
    quajs_native_runtime::capability_manifest_hash(
        &crate::startup::native_startup_renderer_capabilities(),
    )
}
